### ISSUE TITLE
Text width not calculated correctly when strings include ANSI escape codes directly

### ISSUE THREAD

**Issue by milesrout** [jtdaugherty/brick#181]
This is a bit of a weird one, I guess. If you try to use this widget:

```
italics :: String -> Widget n
italics s = str ("\x001b[3m" ++ s ++ "\x001b[23m")
```

the library will calculate the width of this widget as 5 more than it should. The options are probably to:

1. Ignore it
2. Fix it here
3. Fix it in `vty`
4. Support italics all the way down the stack better
5. Support custom attributes using raw escape codes

In favour of 1, I guess it's fairly uncommon to want this. However, I notice that glirc, which uses vty but not brick, would like to support italics but doesn't because it's not supported in vty. That would also suggest that the place to fix this is below the level of brick, then brick can just pick it up through a newer version of vty.

However, I would suggest that it might be useful to allow some kind of custom attribute that just passes through raw ANSI escape codes and counts them as zero-width.

**Comment by jtdaugherty** [comment]
I view this as a case of (1): all output should be sanitized so that it does not contain escapes. In general, Vty and Brick text output functions assume printable characters, which is not actually documented. I'm happy to fix that, though.

With regard to italics, if that ever does get supported, it will be supported in the form of a new Vty attribute style value, not by control sequences in the string. Still, Brick will get that "for free" if and when that happens.

Sorry this bit you!

**Comment by milesrout** [comment]
Do you have an idea of how this could be worked around? Can widgets simply specify their width or is it always calculated based on their content with no opportunity for modification by the widget?

**Comment by emilaxelsson** [comment]
But it only supports color, not italics.

**Comment by jtdaugherty** [comment]
@emilaxelsson I appreciate the idea of contributing `ansiImage` to Brick, but I view that approach as a very context-specific workaround for a situation where output from another terminal program (with escapes) needs to be embedded in a Brick or Vty application. In general that is not something I want to support in the core library, though. While it might be the best approach for some situations, it's not something I want to encourage more broadly, so I think it would be best released as a separate package and it would also be important to include a note saying "be very sure this is what you want."

**Comment by jtdaugherty** [comment]
Also, on the italics note, just FYI: Vty doesn't support italicized content because many terminals don't support it and those that do don't even implement it the same way. Vty tends to support what we can reasonably expect from just about any terminal emulator, with more experimental features like hyperlinking or mouse mode being disabled by default to avoid compatibility issues.

**Comment by emilaxelsson** [comment]
OK @jtdaugherty, I'll keep `ansiImage` as a separate thing then.

**Comment by milesrout** [comment]
@jtdaugherty my ultimate goal is to display variables in an equation editor in italics. the editor is written in Haskell using brick.

Is it in theory possible to extend `vty`'s notion of attributes to include 'wrap it in these escape codes'? I'm not super familiar with how the attribute API is implemented.

**Comment by jtdaugherty** [comment]
> Is it in theory possible to extend vty's notion of attributes to include 'wrap it in these escape codes'? I'm not super familiar with how the attribute API is implemented.

No; the abstraction provided by Vty hides that from the user deliberately.

I did some more investigation, and it looks like it wouldn't be hard to add italic support to Vty in a way that would be harmless on terminals that don't support it, so I'll do that in the next few days.

**Comment by jtdaugherty** [comment]
Italic support is now available in Vty 5.23: https://hackage.haskell.org/package/vty-5.23

Please note that most terminfo descriptors will not include capabilities for italics (`sitm`/`ritm`) so mileage will vary, but if those capabilities are present (e.g. as described at https://alexpearce.me/2014/05/italics-in-iterm2-vim-tmux/) then the feature will work. If the `italic` style is used in the absence of those capabilities, it's harmless.

@milesrout If that addresses your concern, please close this ticket or let me know what else I can do to help. Thanks!

CC @glguy

**Comment by milesrout** [comment]
@jtdaugherty I don't know what to say, man. Thanks, works perfectly!
